cross_validation_mpi.py

```diff
@@ -0,0 +1,189 @@
+        jobs = list(k_splitter.split(X_data, groups=groups))
+        #gets the jobs and splits them into even-sized-lists to be spread across the different cpu's
+        list_jobs_total = [list() for i in range(COMM.size)]
+        for i in range(len(jobs)):
+            j = i % COMM.size
+            list_jobs_total[j].append(jobs[i])
+        jobs = list_jobs_total
+    else:
+        #initalizes variables to pass to other processors, size of X_data must be intialized correctly
+        if (use_metadata):
+            if (col_name is True):
+                X_data = np.empty((47831, 768),dtype='d')
+            else:
+                X_data = np.empty((47831, 768*3), dtype='d')
+        else:
+                X_data = np.empty((shape_data), dtype='d')  
+        y = None
+        jobs = None
+           
+    #get the values from the root processor
+    y = COMM.bcast(y,root=0)
+    jobs = COMM.scatter(jobs, root=0)
+    COMM.Bcast([X_data, MPI.FLOAT], root=0)
+    #run cross-validation on all the different processors
+    results_init = []
+    for job in jobs:
+        train_ind, test_ind = job
+        results_init.append(run_fold(train_ind, test_ind, balance=balance))
+
+    #gather results together
+    results_init = MPI.COMM_WORLD.gather(results_init, root = 0)
+    del jobs
+
+    #compile and save the results
+    if (rank == 0):
+        del X_data
+        del y
+        print("Finished cross validation!")
+        #flatten the total results
+        results_final = [_i for temp in results_init for _i in temp]
+        y_test = list()
+        y_hat = list()
+        #compute the results
+        for hat, test in results_final:
+            y_test += test
+            y_hat += hat
+        accuracy = accuracy_score(y_test, y_hat)
+        f1_macro = f1_score(y_test, y_hat, average='macro')
+        f1_micro = f1_score(y_test, y_hat, average='micro')
+        f1_weighted = f1_score(y_test, y_hat, average='weighted')
+        conf = confusion_matrix(y_test, y_hat) 
+        results = pd.DataFrame(data=np.array([[model_name, accuracy, f1_macro, f1_micro, f1_weighted]]), columns=['classifier', 'accuracy_score', 'f1_score_macro', 'f1_score_micro', 'f1_score_weighted'])
+        print(results)
+        save_results(results, conf) 
+        return results
+ 
+if __name__ == "__main__":   
+    random_state = 32
+    model_name = 'RF_PCA_lower_border'
+    pca = PCA(n_components='mle',random_state=random_state)
+    rf = RandomForestClassifier(random_state=random_state)
+    model = Pipeline(steps=[('pca',pca),('rf',rf)])
+    #model_name = "Naive"
+    #class NaiveModel:
+    #    def fit(self,X_train,y_train):
+    #        self.majority = y_train.value_counts().idxmax()
+    #    def predict(self,X_test):
+    #        y_hat = [self.majority for i in range(len(X_test))]
+    #        return y_hat
+    print("starting "+str(model_name))
+    #model = NaiveModel()   
+    COMM = MPI.COMM_WORLD
+    rank = COMM.rank
```

Comment:
Could all this MPI stuff be hidden inside evaluate_model?

Reply:
I am not sure about this part. It seems that it will go through this part of the code on every processor. I think that if I change this part to be inside evaluate model, it will start a new process that includes all of the processors on each processor. This would require more testing to make sure I am right though.

Reply:
Hmm, interesting. I am not familiar enough with MPI to know. Maybe pass `COMM` into evaluate so its not being referenced as a global variable.